ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs

```diff
@@ -0,0 +1,184 @@
+  test = withOpenMempool bc $ \mempool -> do
+    let genTxs = testTxsToGenTxPairs txs
+        Mempool
+          { addTxs
+          , getSnapshot
+          , zeroIdx
+          } = mempool
+    invalidTxs <- addTxs genTxs
+    MempoolSnapshot
+      { getTxs
+      , getTxsAfter
+      } <- atomically getSnapshot
+    pure $
+      counterexampleMempoolInfo genTxs invalidTxs getTxs $
+      classifyMempoolSize getTxs $
+      getTxs === getTxsAfter zeroIdx
+
+-- | Test that all valid transactions added to a 'Mempool' can be retrieved
+-- afterward.
+--
+-- n.b. In this test, we add a list of arbitrary transactions to an initially
+-- empty 'Mempool', retrieve all of the transactions from the 'Mempool', and
+-- assert that those transactions are /specifically/ equal to the /valid/
+-- transactions that we attempted to add (the invalid ones should have been
+-- dropped).
+prop_Mempool_addTxs_getTxs :: Chain TestBlock
+                           -> [TestTx]
+                           -> Property
+prop_Mempool_addTxs_getTxs bc txs = withMaxSuccess 1000 $
+  tabulate "Transactions" (map constrName txs) $
+  runSimOrThrow test
+ where
+  dropThd :: [(a, b, c)] -> [(a, b)]
+  dropThd = map (\(a, b, _) -> (a, b))
+  --
+  genTxIsValid :: GenTx TestBlock -> Bool
+  genTxIsValid (TestGenTx (ValidTestTx _)) = True
+  genTxIsValid _                           = False
+  --
+  test :: forall s. SimM s Property
+  test = withOpenMempool bc $ \mempool -> do
+    let genTxs = testTxsToGenTxPairs txs
+        Mempool
+          { addTxs
+          , getSnapshot
+          } = mempool
+    invalidTxs <- addTxs genTxs
+    MempoolSnapshot {getTxs} <- atomically getSnapshot
+    pure $
+      counterexampleMempoolInfo genTxs invalidTxs getTxs $
+      classifyMempoolSize getTxs $
+      filter (genTxIsValid . snd) genTxs === dropThd getTxs
+
+-- | Test that invalid transactions can never be added to the 'Mempool'.
+--
+-- n.b. In this test, we add an arbitrary list of transactions to an initially
+-- empty 'Mempool', retrieve all of the transactions from the 'Mempool', and
+-- assert that none of those transactions are invalid.
+prop_Mempool_InvalidTxsNeverAdded :: Chain TestBlock
+                                  -> [TestTx]
+                                  -> Property
+prop_Mempool_InvalidTxsNeverAdded bc txs = withMaxSuccess 1000 $
+  tabulate "Transactions" (map constrName txs) $
+  runSimOrThrow test
+ where
+  genTxIsInvalid :: GenTx TestBlock -> Bool
+  genTxIsInvalid (TestGenTx (InvalidTestTx _)) = True
+  genTxIsInvalid _                             = False
+  --
+  test :: forall s. SimM s Property
+  test = withOpenMempool bc $ \mempool -> do
+    let genTxs = testTxsToGenTxPairs txs
+        Mempool
+          { addTxs
+          , getSnapshot
+          } = mempool
+    invalidTxs <- addTxs genTxs
+    MempoolSnapshot {getTxs} <- atomically getSnapshot
+    pure $
+      counterexampleMempoolInfo genTxs invalidTxs getTxs $
```

Comment:
You can also include this line and the following in the helper. The `tabulate` and `runSimOrThrow` lines too.  Maybe the helper should then just take `MempoolSnapshot _ _ _ -> Property`. I hope this will let you write the properties in just 1-3 lines.